File: Algorithm/day3/p2.cpp
#include <bits/stdc++.h>
using namespace std;

int n;
int ans = 201;
int board[101][101];
bool vis[101][101];
int dist[101][101];

int dx[] = {1, 0, -1, 0};
int dy[] = {0, 1, 0, -1};

int main(){
    cin>>n;
    for(int i=0; i<n; i++){
        for(int j=0; j<n; j++){
            cin>>board[i][j];
        }
    }

    int cnt = 0;
    queue<pair<int, int>> q;
    //각 육지별로 숫자 다르게 부여
    for(int i=0; i<n; i++){
        for(int j=0; j<n; j++){
            if(board[i][j] == 1 && vis[i][j]==0){
                cnt++;
                vis[i][j] = true;
                board[i][j] = cnt;
                q.push({i, j});
                while(!q.empty()){
                    auto cur = q.front();
                    q.pop();
                    for(int dir = 0; dir < 4; dir++){
                        int nx = cur.first + dx[dir];
                        int ny = cur.second + dy[dir];

                        if(nx < 0|| ny < 0|| nx >=n || ny >= n)continue;
                        if(vis[nx][ny] || board[nx][ny]==0)continue;
                        board[nx][ny] = cnt;
                        vis[nx][ny] = 1;
                        q.push({nx, ny});
                    }
                }
            }
        }
    }

    //dist배열을 -1로 초기화
    for(int i=0; i<n; i++) fill(dist[i], dist[i]+n, -1);

    for(int i=0; i<n; i++){
        for(int j=0; j<n; j++){
            if(board[i][j] != 0){
                q.push({i, j});
                dist[i][j] = 0;

                while(!q.empty()){
                    auto cur = q.front();
                    q.pop();

                    for(int dir =0; dir<4; dir++){
                        int nx = cur.first + dx[dir];
                        int ny = cur.second + dy[dir];

                        if(nx < 0|| ny < 0|| nx >=n || ny >= n)continue;
                        if(board[i][j] == board[nx][ny] || dist[nx][ny]>=0)continue;
                        if(board[nx][ny] != 0 && board[nx][ny] != board[i][j]){
                            ans = min(ans, dist[cur.first][cur.second]);
                        }
                        dist[nx][ny] = dist[cur.first][cur.second]+1;
                        q.push({nx, ny});
                    }
                }
                for(int i=0; i<n; i++) fill(dist[i], dist[i]+n, -1);
            }
        }
    }
    cout<<ans;
}

File: Algorithm/day7/P2.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class P2 {
    static int[][] board;
    static int N, M, K;
    static int[][][] dist;
    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, 1, -1};
    static class Pair{
        int x, y, z;
        Pair(int x, int y, int z){
            this.x = x;
            this.y = y;
            this.z = z;
        }
        int getX(){
            return x;
        }
        int getY(){
            return y;
        }
        int getZ(){
            return z;
        }
    }

    static int bfs(int i, int j){
        int ret= 0x7f7f7f7f;
        dist[i][j][0] = 0;
        Queue<Pair> queue = new LinkedList<>();
        queue.offer(new Pair(i, j, 0));
        while (!queue.isEmpty()) {
            Pair cur = queue.poll();
            for(int dir=0; dir<4; dir++){
                int nx = cur.getX() + dx[dir];
                int ny = cur.getY() + dy[dir];
                int z = cur.getZ();

                if(nx < 0|| ny <0 || nx >= N || ny >=K) continue;
                if(board[nx][ny] == 1){
                    if(z+1 > K) continue;
                    if(dist[nx][ny][z+1] != -1) continue;
                    queue.offer(new Pair(nx, ny, z+1));
                    dist[nx][ny][z+1] = dist[cur.getX()][cur.getY()][z] + 1;
                }
                else{
                    if(dist[nx][ny][z] != -1) continue;
                    queue.offer(new Pair(nx, ny, z));
                    dist[nx][ny][z] = dist[cur.getX()][cur.getY()][z] + 1;
                }
            }
        }

        for(int times=0; times<=K; times++){
            if(dist[N-1][K-1][times] == -1)continue;
            ret = Math.min(ret, dist[N-1][K-1][times]);
        }
        if(ret == 0x7f7f7f7f) ret = -1;

        return ret;
    }


    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        K = Integer.parseInt(st.nextToken());
        board = new int[N][M];
        dist = new int[N][M][K];
        for(int i=0; i<N; i++){
            String str = br.readLine();
            for(int j=0; j<M; j++){
                board[i][j] = Character.getNumericValue(str.charAt(j));
            }
        }
//        Arrays.fill(dist, -1); //이렇게 초기화 가능한가?
        /*
        * 이렇게 초기화 해야하는가?
        * */
        for(int i=0; i<N; i++){
            for(int j=0; j<M; j++){
                for(int l=0; l<K; l++){
                    dist[i][j][l] = -1;
                }
            }
        }
        int ans = bfs(0, 0);

    }


}
